rllib/utils/tf_utils.py

```diff
@@ -27,6 +27,51 @@
 tf1, tf, tfv = try_import_tf()
 
 
+@PublicAPI
+def apply_grad_clipping(
+    grads: Union[ModelGradients, List[ModelGradients]],
+    grads_and_vars: Union[ModelGradients, List[ModelGradients]],
+    clip_value: float,
+) -> List[TensorType]:
+    """Applies gradient clipping to a list of gradients.
+
+    Gradient clipping is performd by using the global norm over all
+    gradients in the list. As described by Pascanu et al. (2012) this
+    is the corrct way of gradient clipping.
+
+    As the global norm is infinity all entries of the clipped gradients
+    are set to NaN by TensorFlow to signal that an error occurred. To
+    avoid that this influences the whole training process of an RL
+    algorithm the NaN valus get converted into 0.0 by this funciton, i.e.
+    training can continue, but the gradients at this step have no influence
+    on the weights.
+
+        Args:
+            grads: List of gradients (for each variable ne gradient).
+            grads_and_vars: List of gradient, variable tuples. This has to
+                be updated after clipping and this function does so.
+            clip_value: The clipping ratio for the global norm clipping.
+                Usually this is the `grad_clip` hyperparameter of the algorithm.
+
+        Returns:
+            The updated grads and grads_and_vars after clipping.
+    """
+    grads, _ = tf.clip_by_global_norm(grads, clip_value, name="grad_clip")
+
+    # If the global_norm is inf -> All grads will be NaN. Stabilize this
```

Comment:
I don't agree with this way of doing things. 

if the grads are really small, you could add a really small epsilon to them, but if they're really large, then we should let the NaN happen, or rather, it shouldn't be handled during gradient clipping.

Reply:
I see your point here and mainly agree to it. `NaN` values should be reported. 

- If gradients are really small `tf.clip_by_global_norm()` returns the original gradients and the vanishing gradient problem should be handled with corresponding methods 

- If gradients are exploding the clipping by global norm generally helps, but

- If we have (1) somewhere an `NaN` or `Inf` value in the gradients or (2) the global norm becomes `NaN` or `Inf`, the clipped gradients all become `NaN`. The safe-guard should avoid this by setting the gradients then to zero.

The last case is especially frustrating/costly as then the whole network turns useless and training has to be aborted. Usually an error like [here](https://discuss.ray.io/t/action-masking-error/8347/8) occurs and is uninformative. 

We could for the last case error out with a useful user message instead of silently safe-guarding. We could error out also before clipping, if any gradient is non-finite. I am interested in your opinion.